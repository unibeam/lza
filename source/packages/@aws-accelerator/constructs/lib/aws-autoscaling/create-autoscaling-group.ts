/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { ServiceLinkedRole } from '@aws-accelerator/constructs';

export interface IAutoscalingGroupResource extends cdk.IResource {
  /**
   * The name of the AWS::AutoScaling::AutoScalingGroup resource
   */
  readonly autoscalingGroupName: string;
}

export interface AutoscalingGroupProps {
  readonly name: string;
  readonly minSize: number;
  readonly maxSize: number;
  readonly desiredSize: number;
  readonly launchTemplateVersion: string;
  readonly launchTemplateId: string;
  readonly healthCheckGracePeriod?: number;
  readonly healthCheckType?: string;
  readonly targetGroups?: string[];
  readonly subnets: string[];
  readonly tags?: cdk.CfnTag[];
  readonly lambdaKey?: cdk.aws_kms.IKey;
  readonly cloudWatchLogKmsKey?: cdk.aws_kms.IKey;
  readonly cloudWatchLogRetentionInDays: number;
  readonly maxInstanceLifetime?: number;
  /**
   * Prefix for nag suppression
   */
  readonly nagSuppressionPrefix?: string;
}

export class AutoscalingGroup extends cdk.Resource implements IAutoscalingGroupResource {
  public readonly autoscalingGroupName: string;

  constructor(scope: Construct, id: string, props: AutoscalingGroupProps) {
    super(scope, id);

    // Ensure there is service linked role for autoscaling
    new ServiceLinkedRole(this, 'AutoScalingServiceLinkedRole', {
      environmentEncryptionKmsKey: props.lambdaKey,
      cloudWatchLogKmsKey: props.cloudWatchLogKmsKey,
      cloudWatchLogRetentionInDays: props.cloudWatchLogRetentionInDays,
      awsServiceName: 'autoscaling.amazonaws.com',
      description:
        'Default Service-Linked Role enables access to AWS Services and Resources used or managed by Auto Scaling',
      roleName: 'AWSServiceRoleForAutoScaling',
    });

    const resource = new cdk.aws_autoscaling.CfnAutoScalingGroup(this, 'Resource', {
      minSize: props.minSize.toString(),
      maxSize: props.maxSize.toString(),
      desiredCapacity: props.desiredSize.toString(),
      launchTemplate: {
        version: props.launchTemplateVersion,
        launchTemplateId: props.launchTemplateId,
      },
      healthCheckType: props.healthCheckType!,
      healthCheckGracePeriod: props.healthCheckGracePeriod,
      targetGroupArns: props.targetGroups,
      vpcZoneIdentifier: props.subnets,
      maxInstanceLifetime: props.maxInstanceLifetime,
      tags: props.tags ? this.processTags(props.tags) : undefined,
      // autoScalingGroupName: props.name,
    });
    this.autoscalingGroupName = resource.ref;
  }

  private processTags(tags: cdk.CfnTag[]): cdk.aws_autoscaling.CfnAutoScalingGroup.TagPropertyProperty[] {
    return tags.map(tag => {
      return {
        key: tag.key,
        value: tag.value,
        propagateAtLaunch: true,
      };
    });
  }
}
