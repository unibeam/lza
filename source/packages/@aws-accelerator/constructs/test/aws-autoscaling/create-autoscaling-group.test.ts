/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { AutoscalingGroup } from '../../lib/aws-autoscaling/create-autoscaling-group';
import { snapShotTest } from '../snapshot-test';
import { describe } from '@jest/globals';

const testNamePrefix = 'Construct(AutoscalingGroup): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new AutoscalingGroup(stack, 'Test', {
  name: 'string',
  minSize: 1,
  maxSize: 4,
  desiredSize: 2,
  launchTemplateId: 'string',
  launchTemplateVersion: 'string',
  healthCheckGracePeriod: 300,
  healthCheckType: 'ELB',
  targetGroups: ['string'],
  subnets: ['string'],
  lambdaKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  cloudWatchLogKmsKey: new cdk.aws_kms.Key(stack, 'CustomKeyCloudWatch', {}),
  cloudWatchLogRetentionInDays: 3653,
  maxInstanceLifetime: 86400,
  tags: [{ key: 'key', value: 'value' }],
});

/**
 * GWLB construct test
 */
describe('AutoscalingGroup', () => {
  snapShotTest(testNamePrefix, stack);
});
