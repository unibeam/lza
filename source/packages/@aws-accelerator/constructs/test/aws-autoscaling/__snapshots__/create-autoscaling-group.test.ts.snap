// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoscalingGroup Construct(AutoscalingGroup):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomKeyCloudWatchFB91CD4E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "Test7BFAF513": {
      "Properties": {
        "DesiredCapacity": "2",
        "HealthCheckGracePeriod": 300,
        "HealthCheckType": "ELB",
        "LaunchTemplate": {
          "LaunchTemplateId": "string",
          "Version": "string",
        },
        "MaxSize": "4",
        "MinSize": "1",
        "Tags": [
          {
            "Key": "key",
            "PropagateAtLaunch": true,
            "Value": "value",
          },
        ],
        "TargetGroupARNs": [
          "string",
        ],
        "VPCZoneIdentifier": [
          "string",
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
    },
    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunction48483210": {
      "DependsOn": [
        "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionServiceRoleDefaultPolicy5556EF30",
        "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionServiceRoleD427D69E",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "Custom resource provider to create service linked role",
        "Handler": "index.handler",
        "KmsKeyArn": {
          "Fn::GetAtt": [
            "CustomKey1E6D0D07",
            "Arn",
          ],
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionServiceRoleD427D69E",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionLogGroup4DBD4D73": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CustomKeyCloudWatchFB91CD4E",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunction48483210",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionServiceRoleD427D69E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionServiceRoleDefaultPolicy5556EF30": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "iam:CreateServiceLinkedRole",
                "iam:GetRole",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionServiceRoleDefaultPolicy5556EF30",
        "Roles": [
          {
            "Ref": "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionServiceRoleD427D69E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEvent25455835": {
      "DependsOn": [
        "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEventServiceRoleDefaultPolicy610C6366",
        "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEventServiceRole6926E52E",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Default/Test/AutoScalingServiceLinkedRole/CreateServiceLinkedRoleProvider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunction48483210",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEventServiceRole6926E52E",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEventServiceRole6926E52E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEventServiceRoleDefaultPolicy610C6366": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunction48483210",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunction48483210",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEventServiceRoleDefaultPolicy610C6366",
        "Roles": [
          {
            "Ref": "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEventServiceRole6926E52E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleResourceC5B197D4": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleFunctionLogGroup4DBD4D73",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "TestAutoScalingServiceLinkedRoleCreateServiceLinkedRoleProviderframeworkonEvent25455835",
            "Arn",
          ],
        },
        "description": "Default Service-Linked Role enables access to AWS Services and Resources used or managed by Auto Scaling",
        "roleName": "AWSServiceRoleForAutoScaling",
        "serviceName": "autoscaling.amazonaws.com",
        "uuid": "REPLACED-UUID",
      },
      "Type": "Custom::CreateServiceLinkedRole",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
